Replace resize_dim() with set_sizes_and_strides() in THTensor_(unsqueeze1d) in aten/src/TH/generic/THTensor.cpp

joshim5 opened the pull request:
Summary: Replace resize_dim() with set_sizes_and_strides() in THTensor_(unsqueeze1d) in aten/src/TH/generic/THTensor.cpp, as described in T38058642.

Differential Revision: D13928879

ezyang commented on aten/src/TH/generic/THTensor.cpp:
```diff
@@ -478,17 +478,29 @@ void THTensor_(unsqueeze1d)(THTensor *self, THTensor *src, int dimension)
+  std::vector<int64_t> newSize(/* size */ self->dim()+1);
+  std::vector<int64_t> newStride(/* size */ self->dim()+1);
+
+  for(d = self->dim(); d > dimension; d--)
```
This doesn't look semantics preserving

ezyang replied:
Specifically, we previously began the loop on d at `dim()-1`; and now you are beginning the loop at `dim()`

joshim5 replied:
Previously, we began the loop at self->dim()-1 because self->dim() was expanded before this loop was run. Now that the operation is no longer performed in-place, this is unnecessary. Therefore, we now start at self->dim().

ezyang replied:
You are right!

joshim5 commented on aten/src/TH/generic/THTensor.cpp:
```diff
@@ -478,17 +478,29 @@ void THTensor_(unsqueeze1d)(THTensor *self, THTensor *src, int dimension)
+  for(d = dimension - 1; d >= 0; d--)
+  {
+    newSize[d] = self->size(d-1);
+    newStride[d] = self->stride(d-1);
```
This should read `newSize[d] = self->size(d); newStride[d] = self->stride(d);`. Thanks for catching that. I will update the PR. Anything else?

joshim5 committed:
Replace resize_dim() with set_sizes_and_strides() in THTensor_(unsqueeze1d) in aten/src/TH/generic/THTensor.cpp (#16673)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/16673

Replace resize_dim() with set_sizes_and_strides() in THTensor_(unsqueeze1d) in aten/src/TH/generic/THTensor.cpp, as described in T38058642.

Differential Revision: D13928879

fbshipit-source-id: 088bb68333e9f8555acd9cfb4ec356cb999998ea

ezyang commented:
No, everything else looked good. If you want to be super awesome, a test for the previously failing case would be good. Note that `size(-1)` wraps around and picks up the last element, so I bet accidentally the tests had size(0) and size(n-1) which were the same, which masked the error.

ezyang commented:
Please make sure all CI jobs (both here, and internally) are passing before landing. (We do have a bunch of flaky jobs, however, so if you think a failure is not related, feel free to reach out about it.)
